aiohttp/client_reqrep.py

```diff
@@ -566,20 +567,33 @@ async def write_bytes(
 
                 for chunk in self.body:
                     await writer.write(chunk)  # type: ignore[arg-type]
-        except OSError as exc:
-            if exc.errno is None and isinstance(exc, asyncio.TimeoutError):
-                protocol.set_exception(exc)
-            else:
-                new_exc = ClientOSError(
-                    exc.errno, "Can not write request body for %s" % self.url
+        except OSError as underlying_exc:
+            reraised_exc = underlying_exc
+
+            exc_is_not_timeout = underlying_exc.errno is not None or not isinstance(
+                underlying_exc, asyncio.TimeoutError
+            )
+            if exc_is_not_timeout:
+                reraised_exc = ClientOSError(
+                    underlying_exc.errno,
+                    f"Can not write request body for {self.url !s}",
                 )
-                new_exc.__context__ = exc
```

Comment:
@Dreamsorcerer @bdraco so this place also sets `__context__`, but doesn't that have an effect of making `__cause__` useless?

This https://stackoverflow.com/a/11235957/595220 suggests that it's to be used with the `__traceback__`, though. I wonder if something like [`with_traceback()`](https://docs.python.org/3/library/exceptions.html#BaseException.with_traceback) is needed.

https://docs.python.org/3/library/exceptions.html#exception-context also says
> Setting `__cause__` also implicitly sets the `__suppress_context__` attribute to `True`

and

> The default traceback display code shows these chained exceptions in addition to the traceback for the exception itself. **An explicitly chained exception in `__cause__` is always shown when present. An implicitly chained exception in `__context__` is shown only if `__cause__` is [None](https://docs.python.org/3/library/constants.html#None) and `__suppress_context__` is false.**

Reply:
We set `__context__` before `__cause__` before so changing the order might change the behavior here

Need to dig through https://github.com/python/cpython/blob/dc01b919c721f43ad024ba444a5d19541370e581/Objects/exceptions.c#L360

Reply:
```
/* Steals a reference to cause */
void
PyException_SetCause(PyObject *self, PyObject *cause)
{
    PyBaseExceptionObject *base_self = _PyBaseExceptionObject_cast(self);
    base_self->suppress_context = 1;
    Py_XSETREF(base_self->cause, cause);
}
```

Reply:
Order doesn't matter, but setting `__context__` is likely useless here because it gets suppressed anyways

Reply:
So both `__context__` *and* `__cause__` have been first introduced almost a decade ago in https://github.com/aio-libs/aiohttp/commit/20ee192f6a92e73d6c96289ba5768839ded8a06d. The commit is not coming from a PR and is title `Support exception chaining for ClientRequestError classes`. It doesn't contain a longer description offering any explanation beyond that. The comments discuss the usefulness of the `raise ... from ...` syntax, which is not related to our context.

Reply:
@bdraco @Dreamsorcerer do you think it's fine if I delete this assignment within this PR?

Reply:
Deleted. Marking as resolved.